Mixin: support multiple querier scaled objects

colega opened the pull request:
#### What this PR does

We're experimenting with multiple querier scaled objects, querier-burst and querier-burst-backup. This adapts the mixin to support them on the dashboards.

This is how it looks with `hpa_name: 'keda-hpa-querier(-burst(-backup)?)?'`:

![image](https://user-images.githubusercontent.com/1511481/212946627-337ed63f-0dde-4603-b219-c2569d56eb67.png)

#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- [ ] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

pracucci commented on operations/mimir-mixin/dashboards/reads.libsonnet:
```diff
@@ -212,33 +212,56 @@ local filename = 'mimir-reads.json';
-            'kube_horizontalpodautoscaler_spec_min_replicas{%s, horizontalpodautoscaler="%s"}' % [$.namespaceMatcher(), $._config.autoscaling.querier.hpa_name],
-            'kube_horizontalpodautoscaler_spec_max_replicas{%s, horizontalpodautoscaler="%s"}' % [$.namespaceMatcher(), $._config.autoscaling.querier.hpa_name],
-            'kube_horizontalpodautoscaler_status_current_replicas{%s, horizontalpodautoscaler="%s"}' % [$.namespaceMatcher(), $._config.autoscaling.querier.hpa_name],
+            'sum(kube_horizontalpodautoscaler_spec_min_replicas{%s, horizontalpodautoscaler=~"%s"})' % [$.namespaceMatcher(), $._config.autoscaling.querier.hpa_name],
+            'sum(kube_horizontalpodautoscaler_spec_max_replicas{%s, horizontalpodautoscaler=~"%s"})' % [$.namespaceMatcher(), $._config.autoscaling.querier.hpa_name],
```
It looks weird summing min/max. I understand what you're doing here, but the original idea of displaying min/max was that the current number of replicas will be between that band. That's no more true. What if we do `min(kube_horizontalpodautoscaler_spec_min_replicas)` and `max(kube_horizontalpodautoscaler_spec_max_replicas)` instead? Would be slightly more accurate.

If you accept this suggestion, then shouldn't be a stack anymore.

colega replied:
One of the lines of "current" could be under "min", but the topmost would be between min and max.

I think that having `min(min)` would be also misleading, as that would make that number show half of the real min.

Maybe I should just add another row of panels for `querier-burst` and `querier-burst-backup` and make that feature explicitly enabled?

pracucci replied:
Discussed offline. I'm up to remove min and keep max as `max(max)`.

colega replied:
I've left each one of the max values, as they might be different, and doing the max(max) could hide some of the autoscaler being at max value.

pracucci reviewed (approved):
LGTM. I left a comment about min/max.

colega commented:
I'll hold on merging this because the hpa replicas metric never goes to 0, and I need to find a workaround for that.

colega commented:
@pracucci I've updated the PR and updated the screenshot in the description.

I removed the min replicas completely, as agreed, and rendered the "scaling metric / target" as "desired replicas".

I also added a warning about the "Current" replicas reporting 1 when it's actually 0.

dimitarvdimitrov commented on operations/mimir-mixin/dashboards/reads.libsonnet:
```diff
@@ -212,40 +212,80 @@ local filename = 'mimir-reads.json';
           ], [
-            'Scaling metric',
-            'Target per replica',
+            'Desired for {{ scaledObject }}',
```
nit: the panel name already helps for this
```suggestion
            '{{ scaledObject }}',
```

maybe the same comment for the "Current <>" replicas

dimitarvdimitrov replied:
Also, have you considered stacking the lines for the number of replicas? It'd be easier to see how many replicas there are in total at a glance

colega replied:
I tried stacking at the beginning, but then Grafana lies on the numbers.

If you stack `querier=2, querier-burst=0`, then when you hover on the lines, Grafana says `querier=2, querier-burst=2`.

colega replied:
Done the label change in  https://github.com/grafana/mimir/pull/3962/commits/734ccb49ddaf7be0723bbaf9e6081e5ecbc4c999

dimitarvdimitrov commented on operations/mimir-mixin/dashboards/reads.libsonnet:
```diff
@@ -212,40 +212,80 @@ local filename = 'mimir-reads.json';
-            The minimum, maximum, and current number of querier replicas.
+            The maximum, and current number of querier replicas.
+            Please note that the current number of replicas can still show 1 replica even when scaled to 0.
+            Since HPA never reports 0 replicas, the query will report zero only if the HPA is not active.
```
```suggestion
            Since HPA never reports 0 replicas, the query will report 0 only if the HPA is not active.
```

colega replied:
Done in https://github.com/grafana/mimir/pull/3962/commits/734ccb49ddaf7be0723bbaf9e6081e5ecbc4c999

dimitarvdimitrov reviewed (approved):
LGTM, but I always have nitpicks 🚀  (happy to merge as-is as well, the queries SGTM)

pracucci reviewed (approved):
Good job, LGTM! I just left a minor comment.

colega commented on operations/mimir-mixin/dashboards/reads.libsonnet:
```diff
@@ -212,40 +212,80 @@ local filename = 'mimir-reads.json';
-            $.filterKedaMetricByHPA('keda_metrics_adapter_scaler_metrics_value', $._config.autoscaling.querier.hpa_name),
-            'kube_horizontalpodautoscaler_spec_target_metric{%s, horizontalpodautoscaler="%s"}' % [$.namespaceMatcher(), $._config.autoscaling.querier.hpa_name],
+            |||
+              keda_metrics_adapter_scaler_metrics_value
```
I would prefer leaving it as decimal, this way it's clear that it's just the number coming from the metric, not the actual "desired replicas" that HPA has already decided.
